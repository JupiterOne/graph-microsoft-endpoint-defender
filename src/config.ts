import {
  IntegrationExecutionContext,
  IntegrationInstanceConfig,
  IntegrationStepExecutionContext,
  IntegrationValidationError,
} from '@jupiterone/integration-sdk-core';
import { GraphClient } from './ms-graph/client';
import { IntegrationInstanceConfigFieldMap } from '@jupiterone/integration-sdk-core';

/**
 * A type describing the configuration fields required to execute the
 * integration for a specific account in the data provider.
 *
 * When executing the integration in a development environment, these values may
 * be provided in a `.env` file with environment variables. For example:
 *
 * - `CLIENT_ID=123` becomes `instance.config.clientId = '123'`
 * - `CLIENT_SECRET=abc` becomes `instance.config.clientSecret = 'abc'`
 *
 * Environment variables are NOT used when the integration is executing in a
 * managed environment. For example, in JupiterOne, users configure
 * `instance.config` in a UI.
 */

export const instanceConfigFields: IntegrationInstanceConfigFieldMap = {
  clientId: {
    type: 'string',
  },
  clientSecret: {
    type: 'string',
    mask: true,
  },
  tenant: {
    type: 'string',
  },
  findingSeverity: {
    type: 'string',
  },
  findingsLimit: {
    type: 'string',
  },
};

/**
 * Properties provided by the `IntegrationInstance.config`. This reflects the
 * same properties defined by `instanceConfigFields`.
 */
export interface IntegrationConfig extends IntegrationInstanceConfig {
  /**
   * The Azure application client ID used to identify the program as the
   * registered app. This will be the same for all tenants since this is a
   * multi-tenant application.
   */
  clientId: string;

  /**
   * The Azuser App Registration client secret used to authenticate requests.
   */
  clientSecret: string;

  /**
   * The target directory/tenant ID, identified during the admin consent OAuth
   * flow.
   */
  tenant: string;
}

/**
 * An `IntegrationStepExecutionContext` typed for this integration's
 * `IntegrationInstanceConfig`.
 */
export type IntegrationStepContext =
  IntegrationStepExecutionContext<IntegrationConfig>;

export function validateExecutionConfig(
  executionContext: IntegrationExecutionContext<IntegrationConfig>,
): void {
  const { clientId, clientSecret, tenant } = executionContext.instance.config;

  executionContext.logger.info(
    {
      clientId,
      tenantId: tenant,
    },
    'Configured to make Microsoft Graph API calls to tenantId acting as clientId',
  );

  if (!clientId || !clientSecret || !tenant) {
    throw new IntegrationValidationError(
      'Config requires all of {clientId, clientSecret, tenant}',
    );
  }
}

export async function validateInvocation(
  context: IntegrationExecutionContext<IntegrationConfig>,
) {
  validateExecutionConfig(context);
  const apiClient = new GraphClient(context.logger, context.instance.config);
  await apiClient.verifyAuthentication();
}
