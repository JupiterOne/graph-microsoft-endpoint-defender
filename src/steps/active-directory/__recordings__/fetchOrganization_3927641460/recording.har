{
  "log": {
    "_recordingName": "fetchOrganization",
    "creator": {
      "comment": "persister:JupiterOneIntegationFSPersister",
      "name": "Polly.JS",
      "version": "6.0.5"
    },
    "entries": [
      {
        "_id": "b4391da807354cda8b8ffaab25e22621",
        "_order": 0,
        "cache": {},
        "request": {
          "bodySize": 202,
          "cookies": [],
          "headers": [
            {
              "name": "accept",
              "value": "application/json"
            },
            {
              "name": "content-type",
              "value": "application/x-www-form-urlencoded"
            },
            {
              "name": "accept-encoding",
              "value": "gzip,deflate"
            },
            {
              "name": "user-agent",
              "value": "azsdk-js-identity/1.5.2 core-rest-pipeline/1.9.2 Node/v18.16.0 OS/(arm64-Darwin-23.3.0)"
            },
            {
              "name": "x-ms-client-request-id",
              "value": "0f8eaa0a-5b8e-4ca1-870a-cfeb112df4bb"
            },
            {
              "name": "content-length",
              "value": "202"
            },
            {
              "name": "host",
              "value": "login.microsoftonline.com"
            }
          ],
          "headersSize": 429,
          "httpVersion": "HTTP/1.1",
          "method": "POST",
          "postData": {
            "mimeType": "application/x-www-form-urlencoded",
            "params": [],
            "text": "[REDACTED]"
          },
          "queryString": [],
          "url": "https://login.microsoftonline.com/19ae0f99-6fc6-444b-bd54-97504efc66ad/oauth2/v2.0/token"
        },
        "response": {
          "bodySize": 1772,
          "content": {
            "mimeType": "application/json; charset=utf-8",
            "size": 1772,
            "text": "{\"token_type\":\"Bearer\",\"expires_in\":3599,\"ext_expires_in\":3599,\"access_token\":\"[REDACTED]\"}"
          },
          "cookies": [
            {
              "expires": "2024-04-25T02:58:34.000Z",
              "httpOnly": true,
              "name": "fpc",
              "path": "/",
              "sameSite": "None",
              "secure": true,
              "value": "[REDACTED]"
            },
            {
              "httpOnly": true,
              "name": "x-ms-gateway-slice",
              "path": "/",
              "sameSite": "none",
              "secure": true,
              "value": "[REDACTED]"
            },
            {
              "httpOnly": true,
              "name": "stsservicecookie",
              "path": "/",
              "sameSite": "none",
              "secure": true,
              "value": "[REDACTED]"
            }
          ],
          "headers": [
            {
              "name": "cache-control",
              "value": "no-store, no-cache"
            },
            {
              "name": "pragma",
              "value": "no-cache"
            },
            {
              "name": "content-type",
              "value": "application/json; charset=utf-8"
            },
            {
              "name": "expires",
              "value": "-1"
            },
            {
              "name": "strict-transport-security",
              "value": "max-age=31536000; includeSubDomains"
            },
            {
              "name": "x-content-type-options",
              "value": "nosniff"
            },
            {
              "name": "p3p",
              "value": "CP=\"DSP CUR OTPi IND OTRi ONL FIN\""
            },
            {
              "name": "x-ms-request-id",
              "value": "0c19c030-9434-4bde-8016-7b20b09c8100"
            },
            {
              "name": "x-ms-ests-server",
              "value": "2.1.17573.7 - EUS ProdSlices"
            },
            {
              "name": "x-ms-srs",
              "value": "1.P"
            },
            {
              "name": "x-xss-protection",
              "value": "0"
            },
            {
              "_fromType": "array",
              "name": "set-cookie",
              "value": "[REDACTED]"
            },
            {
              "_fromType": "array",
              "name": "set-cookie",
              "value": "[REDACTED]"
            },
            {
              "_fromType": "array",
              "name": "set-cookie",
              "value": "[REDACTED]"
            },
            {
              "name": "date",
              "value": "Tue, 26 Mar 2024 02:58:34 GMT"
            },
            {
              "name": "content-length",
              "value": "1772"
            }
          ],
          "headersSize": 751,
          "httpVersion": "HTTP/1.1",
          "redirectURL": "",
          "status": 200,
          "statusText": "OK"
        },
        "startedDateTime": "2024-03-26T02:58:33.952Z",
        "time": 405,
        "timings": {
          "blocked": -1,
          "connect": -1,
          "dns": -1,
          "receive": 0,
          "send": 0,
          "ssl": -1,
          "wait": 405
        }
      },
      {
        "_id": "aa33ceaf98a63b8bd52b3986a9ee06cb",
        "_order": 0,
        "cache": {},
        "request": {
          "bodySize": 0,
          "cookies": [],
          "headers": [
            {
              "_fromType": "array",
              "name": "authorization",
              "value": "[REDACTED]"
            },
            {
              "_fromType": "array",
              "name": "client-request-id",
              "value": "7d058fa5-f709-5a7a-4ba6-5c9a20afcffc"
            },
            {
              "_fromType": "array",
              "name": "sdkversion",
              "value": "graph-js/2.2.1 (featureUsage=7)"
            },
            {
              "_fromType": "array",
              "name": "accept",
              "value": "*/*"
            },
            {
              "_fromType": "array",
              "name": "user-agent",
              "value": "node-fetch/1.0 (+https://github.com/bitinn/node-fetch)"
            },
            {
              "_fromType": "array",
              "name": "accept-encoding",
              "value": "gzip,deflate"
            },
            {
              "name": "host",
              "value": "graph.microsoft.com"
            }
          ],
          "headersSize": 2018,
          "httpVersion": "HTTP/1.1",
          "method": "GET",
          "queryString": [],
          "url": "https://graph.microsoft.com/v1.0/organization"
        },
        "response": {
          "bodySize": 974,
          "content": {
            "encoding": "base64",
            "mimeType": "application/json;odata.metadata=minimal;odata.streaming=true;IEEE754Compatible=false;charset=utf-8",
            "size": 974,
            "text": "[\"H4sIAAAAAAAECg==\",\"fFRbT+tGEP4raNpHO9jBDvU+HUROJSo4UCWoUo/yMFmPzZyud63dsduA8t+rdQgEVPVxbzPffJd9gS+uRsGZdlboHwEFTyJ9UOfnrcf+adax9i64RmbadedjPsvOf+5IMD76yfkWLT+jsLOQwIhmIFDfX4BrUJBXSFlTVemi0Yu0KIptuq3LIq0uy6ygRi8WWEMCNRkSqpcotOaOQNnBmAS2Q2BLITw8OUsB1PdNAppldzzXbrDiPy9vSYT8tasJFDyuIAHtCT+Uh3k2z9OsSrN8nReqrFRW/DnhaHAw8hiwpVunD0O9gqk59AZ33zDig9+GnoX8vaWzq+fB09mNFWr99OJsSSMZ13dkBRLgcDcY4d7QEgWPZcOvzq/Ij6wpfLW4NVQfx+rQ/0XCtv3mhBs+wPjaIZtXDpx98NRxoHCLQVY7qz8z934jnn6q36MXS35NFq2sd/0b370LgubA3EGB3lND3lN9i7YdsI2jU9Q5kB48y+7adb1htJpOwd69Y/3/i6fKBkF5gxLEE8mRESH9ZFmj+a8mUNOY8jv94cuPgzjOUnQsbBKQk1nh6moZtWZPWpzfrfiZfh+cIKgXGELUoczzBMQJGlBllmXZPgEMgVtL9YNBG3V4edt5Jz/66iLN52lervNSXZSqqKKvNPa4ZcOyWwnKEEDBUZNI5eQCUPAH29r9HSZDTRxPB7HhTQxTo/GCFlmZZpdlkRbzyyr9pcyK9KLJ5/OqmVeLvID9JoGP6h9bxvT0nkfUuwfvGjYUB46hRy2TvUBBbBt1iN599Cbu7OMzN3JgdzL+JoGRPDdM9dJ1yK+cvE3KMbAwlU3O7puGNV27rhuijPGvCPfWsKUpH0uaYgdK/EAxMDeWhSP5hw17yNyPfGLmJGlLGmfOfvifIAGZLA13aLGlGvab/Wb/LwAAAP//\",\"AwDTx0oH5QQAAA==\"]"
          },
          "cookies": [],
          "headers": [
            {
              "name": "cache-control",
              "value": "no-cache"
            },
            {
              "name": "transfer-encoding",
              "value": "chunked"
            },
            {
              "name": "content-type",
              "value": "application/json;odata.metadata=minimal;odata.streaming=true;IEEE754Compatible=false;charset=utf-8"
            },
            {
              "name": "content-encoding",
              "value": "gzip"
            },
            {
              "name": "vary",
              "value": "Accept-Encoding"
            },
            {
              "name": "strict-transport-security",
              "value": "max-age=31536000"
            },
            {
              "name": "request-id",
              "value": "29b6b72a-6d97-44b4-b608-d0718da8c36e"
            },
            {
              "name": "client-request-id",
              "value": "7d058fa5-f709-5a7a-4ba6-5c9a20afcffc"
            },
            {
              "name": "x-ms-ags-diagnostic",
              "value": "{\"ServerInfo\":{\"DataCenter\":\"Brazil South\",\"Slice\":\"E\",\"Ring\":\"3\",\"ScaleUnit\":\"002\",\"RoleInstance\":\"CP1PEPF00005BC3\"}}"
            },
            {
              "name": "x-ms-resource-unit",
              "value": "1"
            },
            {
              "name": "odata-version",
              "value": "4.0"
            },
            {
              "name": "date",
              "value": "Tue, 26 Mar 2024 02:58:34 GMT"
            },
            {
              "name": "connection",
              "value": "close"
            }
          ],
          "headersSize": 608,
          "httpVersion": "HTTP/1.1",
          "redirectURL": "",
          "status": 200,
          "statusText": "OK"
        },
        "startedDateTime": "2024-03-26T02:58:34.369Z",
        "time": 362,
        "timings": {
          "blocked": -1,
          "connect": -1,
          "dns": -1,
          "receive": 0,
          "send": 0,
          "ssl": -1,
          "wait": 362
        }
      }
    ],
    "pages": [],
    "version": "1.2"
  }
}
