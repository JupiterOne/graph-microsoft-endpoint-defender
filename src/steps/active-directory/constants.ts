import {
  RelationshipClass,
  RelationshipDirection,
} from '@jupiterone/integration-sdk-core';

export const steps: Record<string, string> = {
  FETCH_ACCOUNT: 'Fetch account',
  FETCH_MACHINE: 'Fetch machine',
  MACHINE_DEVICE_RELATIONSHIP: 'Map machine to device',
  FETCH_USERS: 'Fetch users',
  FETCH_FINDINGS: 'Fetch findings',
  FINDING_VULNERABILITY_RELATIONSHIP: 'Map finding to vulnerability',
};

export const entities = {
  ACCOUNT: {
    resourceName: 'Account',
    _type: 'microsoft_defender_account',
    _class: ['Account'],
  },
  MACHINE: {
    resourceName: 'Machine',
    _type: 'microsoft_defender_machine',
    _class: ['HostAgent'],
  },

  USER: {
    resourceName: 'User',
    _type: 'microsoft_defender_logon_user',
    _class: ['User'],
  },
  FINDING: {
    resourceName: 'Vulnerability',
    _type: 'microsoft_defender_vulnerability',
    _class: ['Finding'],
  },
};

export const TargetEntities = {
  CVE: {
    resourceName: 'CVE',
    _type: 'cve',
    _class: ['Vulnerability'],
  },
  DEVICE: {
    resourceName: 'Device',
    _type: 'user_endPoint',
    _class: ['Device/Host'],
  },
};

export const relationships = {
  ACCOUNT_HAS_MACHINE: {
    _type: 'microsoft_defender_account_has_machine',
    sourceType: entities.ACCOUNT._type,
    _class: RelationshipClass.HAS,
    targetType: entities.MACHINE._type,
  },
  MACHINE_HAS_USER: {
    _type: 'microsoft_defender_machine_has_logon_user',
    sourceType: entities.MACHINE._type,
    _class: RelationshipClass.HAS,
    targetType: entities.USER._type,
  },
  MACHINE_IDENTIFIED_FINDING: {
    _type: 'microsoft_defender_machine_identified_vulnerability',
    sourceType: entities.MACHINE._type,
    _class: RelationshipClass.IDENTIFIED,
    targetType: entities.FINDING._type,
  },
};

export const MappedRelationships = {
  FINDING_IS_CVE_VULNERABILITY: {
    _type: 'microsoft_defender_vulnerability_is_cve',
    sourceType: entities.FINDING._type,
    _class: RelationshipClass.IS,
    targetType: TargetEntities.CVE._type,
    direction: RelationshipDirection.FORWARD,
  },
  MACHINE_MANAGES_DEVICE: {
    _type: 'microsoft_defender_machine_manages_user_endPoint',
    sourceType: entities.MACHINE._type,
    _class: RelationshipClass.MANAGES,
    targetType: TargetEntities.DEVICE._type,
    direction: RelationshipDirection.FORWARD,
  },
};

export const DATA_ACCOUNT_ENTITY = entities.ACCOUNT._type;
export const DATA_MACHINE_ENTITY = entities.MACHINE._type;
export const DATA_FINDING_ENTITY = entities.FINDING._type;
