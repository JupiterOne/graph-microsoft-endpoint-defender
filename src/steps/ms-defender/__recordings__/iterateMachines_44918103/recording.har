{
  "log": {
    "_recordingName": "iterateMachines",
    "creator": {
      "comment": "persister:JupiterOneIntegationFSPersister",
      "name": "Polly.JS",
      "version": "6.0.5"
    },
    "entries": [
      {
        "_id": "b4391da807354cda8b8ffaab25e22621",
        "_order": 0,
        "cache": {},
        "request": {
          "bodySize": 233,
          "cookies": [],
          "headers": [
            {
              "name": "accept",
              "value": "application/json"
            },
            {
              "name": "content-type",
              "value": "application/x-www-form-urlencoded"
            },
            {
              "name": "accept-encoding",
              "value": "gzip,deflate"
            },
            {
              "name": "user-agent",
              "value": "azsdk-js-identity/1.5.2 core-rest-pipeline/1.9.2 Node/v18.16.0 OS/(arm64-Darwin-23.3.0)"
            },
            {
              "name": "x-ms-client-request-id",
              "value": "51365985-ca8a-4374-8175-70c8023e3377"
            },
            {
              "name": "content-length",
              "value": "233"
            },
            {
              "name": "host",
              "value": "login.microsoftonline.com"
            }
          ],
          "headersSize": 429,
          "httpVersion": "HTTP/1.1",
          "method": "POST",
          "postData": {
            "mimeType": "application/x-www-form-urlencoded",
            "params": [],
            "text": "[REDACTED]"
          },
          "queryString": [],
          "url": "https://login.microsoftonline.com/19ae0f99-6fc6-444b-bd54-97504efc66ad/oauth2/v2.0/token"
        },
        "response": {
          "bodySize": 1547,
          "content": {
            "mimeType": "application/json; charset=utf-8",
            "size": 1547,
            "text": "{\"token_type\":\"Bearer\",\"expires_in\":3599,\"ext_expires_in\":3599,\"access_token\":\"[REDACTED]\"}"
          },
          "cookies": [
            {
              "expires": "2024-04-25T02:55:11.000Z",
              "httpOnly": true,
              "name": "fpc",
              "path": "/",
              "sameSite": "None",
              "secure": true,
              "value": "[REDACTED]"
            },
            {
              "httpOnly": true,
              "name": "x-ms-gateway-slice",
              "path": "/",
              "sameSite": "none",
              "secure": true,
              "value": "[REDACTED]"
            },
            {
              "httpOnly": true,
              "name": "stsservicecookie",
              "path": "/",
              "sameSite": "none",
              "secure": true,
              "value": "[REDACTED]"
            }
          ],
          "headers": [
            {
              "name": "cache-control",
              "value": "no-store, no-cache"
            },
            {
              "name": "pragma",
              "value": "no-cache"
            },
            {
              "name": "content-type",
              "value": "application/json; charset=utf-8"
            },
            {
              "name": "expires",
              "value": "-1"
            },
            {
              "name": "strict-transport-security",
              "value": "max-age=31536000; includeSubDomains"
            },
            {
              "name": "x-content-type-options",
              "value": "nosniff"
            },
            {
              "name": "p3p",
              "value": "CP=\"DSP CUR OTPi IND OTRi ONL FIN\""
            },
            {
              "name": "x-ms-request-id",
              "value": "879d3886-5ea6-43df-b5fd-d50af330a300"
            },
            {
              "name": "x-ms-ests-server",
              "value": "2.1.17573.7 - NCUS ProdSlices"
            },
            {
              "name": "x-ms-srs",
              "value": "1.P"
            },
            {
              "name": "x-xss-protection",
              "value": "0"
            },
            {
              "_fromType": "array",
              "name": "set-cookie",
              "value": "[REDACTED]"
            },
            {
              "_fromType": "array",
              "name": "set-cookie",
              "value": "[REDACTED]"
            },
            {
              "_fromType": "array",
              "name": "set-cookie",
              "value": "[REDACTED]"
            },
            {
              "name": "date",
              "value": "Tue, 26 Mar 2024 02:55:10 GMT"
            },
            {
              "name": "content-length",
              "value": "1547"
            }
          ],
          "headersSize": 752,
          "httpVersion": "HTTP/1.1",
          "redirectURL": "",
          "status": 200,
          "statusText": "OK"
        },
        "startedDateTime": "2024-03-26T02:55:11.100Z",
        "time": 416,
        "timings": {
          "blocked": -1,
          "connect": -1,
          "dns": -1,
          "receive": 0,
          "send": 0,
          "ssl": -1,
          "wait": 416
        }
      },
      {
        "_id": "aa33ceaf98a63b8bd52b3986a9ee06cb",
        "_order": 0,
        "cache": {},
        "request": {
          "bodySize": 0,
          "cookies": [],
          "headers": [
            {
              "_fromType": "array",
              "name": "authorization",
              "value": "[REDACTED]"
            },
            {
              "_fromType": "array",
              "name": "accept",
              "value": "*/*"
            },
            {
              "_fromType": "array",
              "name": "user-agent",
              "value": "node-fetch/1.0 (+https://github.com/bitinn/node-fetch)"
            },
            {
              "_fromType": "array",
              "name": "accept-encoding",
              "value": "gzip,deflate"
            },
            {
              "name": "host",
              "value": "api.securitycenter.microsoft.com"
            }
          ],
          "headersSize": 1712,
          "httpVersion": "HTTP/1.1",
          "method": "GET",
          "queryString": [],
          "url": "https://api.securitycenter.microsoft.com/api/machines"
        },
        "response": {
          "bodySize": 3240,
          "content": {
            "encoding": "base64",
            "mimeType": "application/json; odata.metadata=minimal; odata.streaming=true; charset=utf-8",
            "size": 3240,
            "text": "[\"5Ftdc9u4Ff0vah9DGt8ffKodx0066+xO401nurMPIAjIzEqiSlKOs5n89x5QlmzZcprYShy7nokzAqELgADOuffc64+jvzWV613um1kfzvtRMTrt+3lX7O25eZ13wS/auv/gA562+bT2bdM1sUf3aeqw99dp6F0y8Jdj50/rWehGz0ZnbrIIo+K3j6O6gkEmKJGqVJYR7pQiQUtGvfG8jJ4JraLxkblA8M1paMehejXrmwtzr2BgtphMno3q7pemxzRqNzlczCe1d33dzEZFdJMupMcvzv1kUQV84aIppIYOnf4ZXJe6Lg1h6vMFVnM46167KeY5entM86aex/ez+t2YnYbzyZmIdPqeLsbvxuO8PM/rtPyZm+R+0iwqN5/ns9BjvrFuu/5NCDA+YoSJjNCMyRNqCiELqXJljKHi3+g5cV/Ysel+mbg+Nu0UNn8tF7N+ga833dvQprWsVoFubeND1zUt+p0rgU5nqy6YS54a0qCv5vtV1aIjulGSqxz/Lh69OF+uaqOLZjn2KWec5DLZcGO89PXgIzRTPOMsp9Yokadda7qDRT1Z79RpcJP+9E3v+kUa9NXM+b4+C+hYhbPah7fL4zF6jTW6CZrb0vm/t81innabXPm83J/ltrV198cb37Rpw143s2QunM+bbtGGn8JZmKD5p+Y9Wutu31X/aHAW1zNyrjochr48TtPlcT1xY8zxt9/T3GKYVaHdP1vP/HXTv1nM503bwxSWOSsb11b1bLzu8fOyafm4229xBfrge8wJ01EiK+t0SqZuhrdYHXxA46+zP2bN+9nV1rW1y2f1atNwn4Z7tPoMC1c2EYu43FxCyCHft/sHB3wf5vsP8zSJn/vT0KbJz0M73Bg3uRxvPvr0DJf0ivEYDCkKRqqCuxiLGKwrypK7Yb67HowyjdOIA7XFOlaz/LnzUopiR3ZxOM6mxxc4Nyo+4lM6RyPmnaFWkExIFTMRpMpKQ1ymTCkir3jlGMfsB8jAZT2rcbrwtf0/0/HAKQ9ds2h9GGztdYuy8209T6DW7VHLo7GVzxRjMhPOkAx3pMysKjmLZVRVRfZWBoar0+29o7hee/OLgbq94zVYP18i3t5Z3fYLN1kh9R5wD/O4OvIwly8ZfPRpODnpNXBmBaPCCerLyD0VZeV4RYGGXpZaBO+UNkYMm/Eg8D4GCjazrD6bZn3o+nzRZYnPWjehGYgvf0ezsc8SxI+XVyTDi8zO+Br18ZKuAT3PKMuIOmGs4KKQItdGKGZvAv2VjkIXhOVca6Vl6viNgZ4yk1OZM7Yigm1ozyVgnKOfyRnVmNRWtMde5vxHR/oL0n9qUL+xjRt4LxihZN8QcvQikfB98F4QQosl3gP+49IVu8otOxnrkcO91NZ5YWUWaGCZiCXNSq1CZqTTRAU8tSX24XvB/djPr2PWVvB/+aFs62pFARduT7d3DRQzQYBMWhMOSpNcW2EZFnNfbpDSME+VZ6UxSsHFMpwyyRkpqZZCGKqMikqm9/Ywrj/iAtf+Z1GG3bAC3H+qT6gtGC2kzoXRNykBQQJPQQLjBdOFtDnlCh73DUo4DGXtkqP4lb7/yuPb4vyvwIRamE3Pt3KCyKmm4AWF/5MDs40TBKGIAL6AE/Yfzvd/ioyw9M8rW0Q/uOdUFzHS5E5ucANhh/boOdVHR/QFnt2NGyhCQZo89Jt+NBPsSB4cHBIj72x+uRTBiiiHpZRlUS3tXVvKDsZCzLRx9jdGuCS3w4P7rmaTSKsB2LYT6T3GeuRECr0HwWVpM625z4QXZWZEdEBO7l1JrdchKRSPhEjXDJIJLhXTPCk+hFhJwKZYxn0pFKEej9oHyiz1zPtKO4rIkkViuSi9szwEFgYv/2EotJ51vZv5kNEdcajOmD2htGCmwL0lhCs6sOMNCQ00qk4IKYgohMml1ohlvieNDuH9Z2kUAArBjMl0ELbRKCcKXbBIPP+8iva9ePQ4VPViiulsyGhPkUqvcMKwkbdxwhFext0I9ILhrnFChL3dj/XIOaEilbBaiCy6JHgRbzNTyZgZzhyPVTRUJk/jkXDCJSSCFLSkhmkquQU5EE12EVfpEoJbKIm2jFdepyAKbqANXppoEIkGLqgKzqej9iCaG9S0zE8rT+iQWULSKKWMMJ0bQhpRGYWQhnhJFuhNmJaS0q1hE3Ir6oTKQgiobjm6cc5u4P2/6lnVvO/ehBbpEORk0uv+ygiK0ZfpW1sCqKRWf05PQ96FJd3NIHYyW7InwBwjFToRDsailiRHZw38Q+uz0YPlT17W41PM5wki/20SvgmaBqVsVjKpM1HGMnNeVxk0C6MDd9aI76npvHp9kj0/3irjrASc6xr+5UXDzt3X17QlNdZIGyxXhPOgSyFU5QinyIeGUHHJSvx6OFgZN7M/3aTZebYWYbSQuJpWK3Mb9Axp3esdN0T8lfLylXCz+toWvPlf2VotkKiFhk8pJo8DsN3H/CGTtRc53A2wWWZ6n5qEv9rEDaePgJmEYc+ZMndXaC6ytchCCTNoKMwXMHhTDtrFYI/cw/RwwJL2nCGtqpCyheBgeWUyazmvAiGRyfTevpeH+fXZ2gvs2wHMe2ZdBYYTqE5BCQ5S2DYoR2iUlgoU6CjBgpRDUcPDeI9vj2PbTLtT1yIanQLUdgn4gwsJLzIVHxir+HaJfuVrslVHI2/6mt+mPMdgi29TFSTkgsG5RHXOrfnaH7M65/8L8NMmbgC+IkIeHB4KIp7fVzNXDNVlZTUgfgXtyyeDux/skQO+Q3mOVwK6ZGQ0Ewy/bBnx0SMnSZSKfEC4Hxfwb6DgDqAfKVhtqlIZLq1hQTjE2nD5SYwxaQaIdzhh0Q4a1YMIBwn4X+0a8uG66wKKsqS5oBLn+jM+vi44OiIHChq0QzCw4eN/G8hPF/g2yE9puGVF5jLf+sh9/CcqJScn/xYYfvGCi0H1u4+KfBXzQ4D+NVjcAvr3HO2Rg77VxkZCqswoVOoIIlGkI1SZecuZ5wrC8lBs8uOC/hr/dgD2UaBYFNWoCHhKIjnnKEtN7r4JJdMaJTnIt8byQQvvj74R3JsCjju1ObKx1m6py1xW6iMUQPUmL1AnCdoz6nt5+Mlzvw3u15IOyW8rvnk85fdPGOzTHm7A71LRoUdWseT83wfs2aWiQ2NxYXD3gz1yrIdmE5SXqMA0tkRBpkTOUBibRQs/H2XohMv0dws/Lta/PV7DH+Z5V/H+90//BQAA//8DAA==\"]"
          },
          "cookies": [],
          "headers": [
            {
              "name": "date",
              "value": "Tue, 26 Mar 2024 02:55:12 GMT"
            },
            {
              "name": "content-type",
              "value": "application/json; odata.metadata=minimal; odata.streaming=true; charset=utf-8"
            },
            {
              "name": "transfer-encoding",
              "value": "chunked"
            },
            {
              "name": "connection",
              "value": "close"
            },
            {
              "name": "content-encoding",
              "value": "deflate"
            },
            {
              "name": "vary",
              "value": "Accept-Encoding"
            },
            {
              "name": "odata-version",
              "value": "4.0"
            },
            {
              "name": "strict-transport-security",
              "value": "max-age=31536000; includeSubDomains"
            }
          ],
          "headersSize": 313,
          "httpVersion": "HTTP/1.1",
          "redirectURL": "",
          "status": 200,
          "statusText": "OK"
        },
        "startedDateTime": "2024-03-26T02:55:11.524Z",
        "time": 558,
        "timings": {
          "blocked": -1,
          "connect": -1,
          "dns": -1,
          "receive": 0,
          "send": 0,
          "ssl": -1,
          "wait": 558
        }
      }
    ],
    "pages": [],
    "version": "1.2"
  }
}
