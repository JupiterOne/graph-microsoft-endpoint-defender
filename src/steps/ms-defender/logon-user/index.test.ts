import { executeStepWithDependencies } from '@jupiterone/integration-sdk-testing';
import { Steps } from '../../../constants';
import { Recording, setupProjectRecording } from '../../../../test/recording';
import { buildStepTestConfigForStep } from '../../../../test/config';

let recording: Recording;
afterEach(async () => {
  await recording.stop();
});

test.skip('#fetchLogonUsers', async () => {
  recording = setupProjectRecording({
    directory: __dirname,
    name: '#fetchLogonUsers',
  });

  const stepConfig = buildStepTestConfigForStep(Steps.FETCH_LOGON_USERS.id);
  const stepResult = await executeStepWithDependencies(stepConfig);
  expect(stepResult).toMatchStepMetadata(stepConfig);
  expect(stepResult).toMatchSnapshot();
});
