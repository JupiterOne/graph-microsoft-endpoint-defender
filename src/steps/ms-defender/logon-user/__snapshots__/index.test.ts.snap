// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#fetchLogonUsers 1`] = `
Object {
  "collectedData": Object {},
  "collectedEntities": Array [
    Object {
      "_class": Array [
        "User",
      ],
      "_key": "microsoft_defender_logon_user:testmachine1\\\\administrator1:f39177bfecabebc12576671381d1f95c68492ae3",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "accountDomain": "testmachine1",
            "accountName": "administrator1",
            "accountSid": null,
            "firstSeen": "2022-09-13T09:00:15Z",
            "id": "testmachine1\\\\administrator1",
            "isDomainAdmin": true,
            "isOnlyNetworkUser": null,
            "lastSeen": "2022-09-13T09:00:22Z",
            "leastPrevalentMachineId": null,
            "logOnMachinesCount": null,
            "logonTypes": "Network, RemoteInteractive",
            "mostPrevalentMachineId": null,
          },
        },
      ],
      "_type": "microsoft_defender_logon_user",
      "active": true,
      "createdOn": undefined,
      "displayName": "administrator1",
      "domain": "testmachine1",
      "firstSeenOn": 1663059615000,
      "id": "testmachine1\\\\administrator1",
      "lastSeenOn": 1663059622000,
      "logonTypes": "Network, RemoteInteractive",
      "name": "administrator1",
      "username": "administrator1",
    },
  ],
  "collectedRelationships": Array [
    Object {
      "_class": "HAS",
      "_fromEntityKey": "microsoft_defender_machine:f39177bfecabebc12576671381d1f95c68492ae3",
      "_key": "microsoft_defender_machine:f39177bfecabebc12576671381d1f95c68492ae3|has|microsoft_defender_logon_user:testmachine1\\\\administrator1:f39177bfecabebc12576671381d1f95c68492ae3",
      "_toEntityKey": "microsoft_defender_logon_user:testmachine1\\\\administrator1:f39177bfecabebc12576671381d1f95c68492ae3",
      "_type": "microsoft_defender_machine_has_logon_user",
      "displayName": "HAS",
    },
  ],
  "encounteredEntityKeys": Set {
    "microsoft_defender_logon_user:testmachine1\\\\administrator1:f39177bfecabebc12576671381d1f95c68492ae3",
    "microsoft_defender_account:local-integration-instance",
    "microsoft_defender_machine:595967ce83eb04bfddedaa7f29c8d10142a5bd30",
    "microsoft_defender_machine:614f81ab67349d2e017f7f8bee5764e1f4a6397d",
    "microsoft_defender_machine:f39177bfecabebc12576671381d1f95c68492ae3",
  },
  "encounteredTypes": Array [
    "microsoft_defender_account",
    "microsoft_defender_machine",
    "microsoft_defender_account_has_machine",
    "microsoft_defender_logon_user",
    "microsoft_defender_machine_has_logon_user",
  ],
}
`;
